pkg/image/origins.go

```diff
@@ -0,0 +1,306 @@
+		"rke-tools":                                               "https://github.com/rancher/rke-tools",
+		"rke2-cloud-provider":                                     "https://github.com/rancher/rke2",
+		"rke2-runtime":                                            "https://github.com/rancher/rke2",
+		"rke2-upgrade":                                            "https://github.com/rancher/rke2-upgrade",
+		"security-scan":                                           "https://github.com/rancher/security-scan",
+		"shell":                                                   "https://github.com/rancher/shell",
+		"system-agent":                                            "https://github.com/rancher/system-agent",
+		"system-agent-installer-k3s":                              "https://github.com/rancher/system-agent-installer-k3s",
+		"system-agent-installer-rke2":                             "https://github.com/rancher/system-agent-installer-rke2",
+		"system-upgrade-controller":                               "https://github.com/rancher/system-upgrade-controller",
+		"tekton-utils":                                            "https://github.com/rancher/build-tekton",
+		"thanosio-thanos":                                         "https://github.com/thanos-io/thanos",
+		"ui-plugin-operator":                                      "https://github.com/rancher/ui-plugin-operator",
+		"weave-kube":                                              "https://github.com/weaveworks-experiments/weave-kube",
+		"weave-npc":                                               "https://github.com/weaveworks/weave",
+		"webhook-receiver":                                        "https://github.com/rancher/webhook-receiver",
+		"windows_exporter-package":                                "https://github.com/rancher/windows_exporter-package",
+		"wins":                                                    "https://github.com/rancher/wins",
+		"wmi_exporter-package":                                    "https://github.com/rancher/wmi_exporter-package",
+	}
+}
+
+func GenerateImageOrigins(linuxImagesFromArgs, targetImages, targetWindowsImages []string) error {
+	logrus.Infof("building rancher-image-origins.txt")
+
+	allImages := GetAllUniqueImages(linuxImagesFromArgs, targetImages, targetWindowsImages)
+
+	// if any image does not have a source
+	// declared, or if a net-new image has been
+	// added via KDM or charts and is not
+	// included in the OriginMap
+	// then we cannot continue as the
+	// artifact would be incomplete.
+	if UnknownImages := GatherUnknownImages(allImages); len(UnknownImages) > 0 {
+		return fmt.Errorf("[ERROR] not all images have a source code origin defined. Please provide origin URL's within rancher/pkg/image/origins.go for the following images: \n%s", UnknownImages)
+	}
+
+	// format file
+	fileContents := ""
+	for k, v := range OriginMap {
+		fileContents = fileContents + k + " " + v + "\n"
```

Comment:
QQ: how long does this whole script take to run?

Reply:
If you have the image needed for dapper already cached then it takes about 1 minute, most of that time is spent cloning the charts repo and unpacking it so we can get at the images. If you don't have it cached it might take 2-3 minutes, but again that's largely due to dapper setting up and cloning the repositories which can't really be avoided.